src/tir/transforms/profile_instrumentation.cc

```diff
@@ -0,0 +1,287 @@
+        height = std::max(height, TraverseLoop(n->else_case, parent_depth, has_parallel));
+      return height;
+    } else if (stmt->IsInstance<ForNode>()) {
+      const ForNode* f = stmt.as<ForNode>();
+      LoopInfo loop_info(start_id, parent_depth + 1);
+      start_id++;
+      bool parent_parallel = false;
+      if (has_parallel) {
+        loop_info.has_parallel = true;
+        parent_parallel = true;
+      } else if (f->kind == ForKind::kParallel) {
+        // has_parallel for the current loop is being set to 'false' since the
+        // intrinsic is added outside of the loop. The instrumentation isn't
+        // allowed for the subsequent nested loops.
+        loop_info.has_parallel = false;
+        parent_parallel = true;
+      }
+      unsigned height = TraverseLoop(f->body, parent_depth + 1, parent_parallel);
+      loop_info.height = height;
+      loops[f] = loop_info;
+      return height + 1;
+    } else if (stmt->IsInstance<LetStmtNode>()) {
+      const LetStmtNode* n = stmt.as<LetStmtNode>();
+      return TraverseLoop(n->body, parent_depth, has_parallel);
+    } else if (stmt->IsInstance<AttrStmtNode>()) {
+      const AttrStmtNode* n = stmt.as<AttrStmtNode>();
+      return TraverseLoop(n->body, parent_depth, has_parallel);
+    } else if (stmt->IsInstance<AllocateNode>()) {
+      const AllocateNode* n = stmt.as<AllocateNode>();
+      return TraverseLoop(n->body, parent_depth, has_parallel);
+    } else {
+      return 0;  // inner-most loop
+    }
+  }
+
+ private:
+  LoopInfoMap loops;
+};
+
+class InstrumentIntrin : public StmtMutator {
+ public:
+  InstrumentIntrin(int32_t max_depth, int32_t min_height, bool instr_siblings)
+      : max_instr_depth_(max_depth),
+        min_instr_height_(min_height),
+        instr_siblings_(instr_siblings) {}
+
+  void GetLoopInfo(PrimFuncNode* op) {
+    LoopAnalyzer analzer;
+    loops_ = std::move(analzer.Analyze(op->body));
+  }
+
+  Stmt VisitStmt_(const SeqStmtNode* op) final {
+    Stmt stmt = StmtMutator::VisitStmt_(op);
+    return SeqStmt::Flatten(stmt);
+  }
+
+  Stmt VisitStmt_(const ForNode* op) final {
+    Stmt stmt = StmtMutator::VisitStmt_(op);
+    if (loops_.count(op) < 1) return stmt;
+
+    LoopInfo loop_info = loops_[op];
+
+    if (loop_info.has_parallel) {
+      return stmt;
+    }
+
+    // Exclude inner-most loops from instrumentation. The inner-most loop has
+    // height '0' and it increases as we move outward in the loop nest.
+    if (loop_info.height < min_instr_height_) {
+      return stmt;
+    }
+
+    // Only instrument loops with a sibling
+    if (instr_siblings_ && !loop_info.has_siblings) {
+      return stmt;
+    }
+
+    // If instr_siblings_ is set, ignore max depth for instrumentation
+    if (!instr_siblings_ && loop_info.depth > max_instr_depth_) {
+      return stmt;
+    }
+    PrimExpr id = static_cast<int32_t>(loop_info.id);
+    PrimExpr call = Call(DataType::Handle(), builtin::profile_intrinsic(), {id});
```

Comment:
Sure! It makes sense to have two separate intrinsics to mark the start and end of the code segments. This would definitely be more useful for some metrics.